[RISCV] Enable rematerialization for scalar loads

lukel97 opened the pull request:
In some workloads we see an argument passed on the stack where it is loaded, only for it to be immediately spilled to a different slot on the stack and then reloaded from that spill slot later on.

We can avoid the unnecessary spill by marking loads as rematerializable and just directly loading from where the argument was originally passed on the stack. TargetTransformInfo::isReMaterializableImpl checks to make sure that any loads are `MI.isDereferenceableInvariantLoad()`, so we should be able to move the load down to the remat site. 

This gives a 14.8% reduction in spills in 544.nab_r on rva23u64 -O3, and a few other smaller reductions on llvm-test-suite. I didn't find any benchmarks where the number of spills/reloads increased.

Related: #165761

lukel97 committed:
Precommit tests

lukel97 committed:
[RISCV] Enable rematerialization for scalar loads

preames commented on llvm/test/CodeGen/RISCV/remat.ll:
```diff
@@ -201,3 +201,95 @@ for.end:                                          ; preds = %for.inc, %entry
+  store i32 %stack1, ptr %q
+  tail call void asm sideeffect "", "~{x1},~{x3},~{x4},~{x5},~{x6},~{x7},~{x8},~{x9},~{x10},~{x11},~{x12},~{x13},~{x14},~{x15},~{x16},~{x17},~{x18},~{x19},~{x20},~{x21},~{x22},~{x23},~{x24},~{x25},~{x26},~{x27},~{x28},~{x29},~{x30},~{x31}"()
+  %a = icmp eq i32 %0, 0
+  br i1 %a, label %truebb, label %falsebb
```
I think you can simplify the test.  Is the conditional block doing anything in particularly here?  If you just had the inline assembly followed by a unconditional use of %stack0 and %stack1, I think that still remats?  Or even, just return one of %stack0 or %stack1?

lukel97 replied:
Yup, was able to simplify it to just a volatile store after the asm without branching. For some reason, the float stack arguments also require a use before the asm in order to force the initial load off the stack

preames reviewed (commented):
Summarizing some offline conversation with Craig.  We happened to chat about this idea yesterday.  

Cases this might kick in:
- Arguments - will be frame indices during regalloc, and rewritten afterwards, so no concerns for e.g. dynamic alloc offset calculations.
- Globals, particularly pc-relative.  Remating the LD without the whole sequence might break the assumption of the AUIPC offset?

Craig, can you expand on the later just a bit?  As long as we don't move the AUIPC, isn't the low offset fixed?  It's the pc relative offset from the AUIPC to the global, not from the LD to the global isn't it?

Luke, could you add a test case which does a constant global load before the asm block, and then a use after?  We should be able to remat that, but it's the case we were a bit unclear on so having specific testing is good.

topperc commented:
> Craig, can you expand on the later just a bit? As long as we don't move the AUIPC, isn't the low offset fixed? It's the pc relative offset from the AUIPC to the global, not from the LD to the global isn't it?

You're right. Remating the load should be fine. It has a reference to a symbol on the auipc. We can remat that since we can have multiple instructions referring to the same auipc. It's auipc that we can't rematerialize since that would need a new symbol.

bababuck commented:
Any reason this wouldn't apply to vector as well? Though I don't recall ever seeing any cases in the wild that would benefit from this though...

topperc commented:
> Any reason this wouldn't apply to vector as well? Though I don't recall ever seeing any cases in the wild that would benefit from this though...

I don't think the generic infrastructure can identify an invariant vector load from stack since the frame index operand won't be part of the load.

Maybe it's possible for other types of invariant loads?

lukel97 commented on llvm/test/CodeGen/RISCV/remat.ll:
```diff
@@ -201,3 +201,95 @@ for.end:                                          ; preds = %for.inc, %entry
+  ret i32 %b
+falsebb:
+  ret i32 0
+}
```
Done in the new precommitted test in bc86a8f56a48de75c969bee4dfd3e8a8940e0209

lukel97 commented:
Apologies for the force push, I precomitted 630f43a305b9e48d6c14ecb53a50e511402e2efb and rebased on top of it so we can test i64 ld remat.

> Luke, could you add a test case which does a constant global load before the asm block, and then a use after? We should be able to remat that, but it's the case we were a bit unclear on so having specific testing is good.

I've added a test for a constant global load, but it looks like we can't rematerialize global loads yet because even though the load is invariant, the upper bits of the address won't be live at the point of rematerialization. I think in order to remat this we either need to extend the live range or allow rematerializing multiple instructions.

lukel97 commented on llvm/test/CodeGen/RISCV/remat.ll:
```diff
@@ -200,3 +200,168 @@ for.end:                                          ; preds = %for.inc, %entry
+; CHECK-NEXT:    addi sp, sp, 208
+; CHECK-NEXT:    ret
+entry:
+  ; Force loading the stack arguments to create their live interval
```
If we don't add a use before the asm, then the machine scheduler will sink the loads of the floats from the stack after the asm call:

```
0B	bb.0.entry:
16B	  %16:fpr64 = FLD %fixed-stack.1, 0 :: (load (s64) from %fixed-stack.1)
32B	  %17:fpr32 = FLW %fixed-stack.2, 0 :: (load (s32) from %fixed-stack.2, align 16)
48B	  %18:fpr16 = FLH %fixed-stack.3, 0 :: (load (s16) from %fixed-stack.3, align 8)
64B	  %19:fpr16 = FLH %fixed-stack.4, 0 :: (load (s16) from %fixed-stack.4, align 16)
80B	  %20:gpr = LD %fixed-stack.5, 0 :: (load (s64) from %fixed-stack.5)
96B	  %21:gpr = LD %fixed-stack.6, 0 :: (load (s64) from %fixed-stack.6, align 16)
112B	  %22:gpr = LD %fixed-stack.7, 0 :: (load (s64) from %fixed-stack.7)
128B	  %23:gpr = LD %fixed-stack.0, 0 :: (load (s64) from %fixed-stack.0, align 16)
144B	  %24:gpr = LD %fixed-stack.8, 0 :: (load (s64) from %fixed-stack.8, align 16)
160B	  INLINEASM &"" ...
176B	  SB %24:gpr, %23:gpr, 0 :: (volatile store (s8) into %ir.p)
192B	  SH %22:gpr, %23:gpr, 0 :: (volatile store (s16) into %ir.p)
208B	  SW %21:gpr, %23:gpr, 0 :: (volatile store (s32) into %ir.p)
224B	  SD %20:gpr, %23:gpr, 0 :: (volatile store (s64) into %ir.p)
240B	  FSH %19:fpr16, %23:gpr, 0 :: (volatile store (s16) into %ir.p)
256B	  FSH %18:fpr16, %23:gpr, 0 :: (volatile store (s16) into %ir.p)
272B	  FSW %17:fpr32, %23:gpr, 0 :: (volatile store (s32) into %ir.p)
288B	  FSD %16:fpr64, %23:gpr, 0 :: (volatile store (s64) into %ir.p)
304B	  PseudoRET
```

```
0B	bb.0.entry:
80B	  %20:gpr = LD %fixed-stack.5, 0 :: (load (s64) from %fixed-stack.5)
128B	  %23:gpr = LD %fixed-stack.0, 0 :: (load (s64) from %fixed-stack.0, align 16)
144B	  %24:gpr = LD %fixed-stack.8, 0 :: (load (s64) from %fixed-stack.8, align 16)
148B	  %22:gpr = LD %fixed-stack.7, 0 :: (load (s64) from %fixed-stack.7)
152B	  %21:gpr = LD %fixed-stack.6, 0 :: (load (s64) from %fixed-stack.6, align 16)
160B	  INLINEASM ...
176B	  SB %24:gpr, %23:gpr, 0 :: (volatile store (s8) into %ir.p)
192B	  SH %22:gpr, %23:gpr, 0 :: (volatile store (s16) into %ir.p)
208B	  SW %21:gpr, %23:gpr, 0 :: (volatile store (s32) into %ir.p)
224B	  SD %20:gpr, %23:gpr, 0 :: (volatile store (s64) into %ir.p)
228B	  %19:fpr16 = FLH %fixed-stack.4, 0 :: (load (s16) from %fixed-stack.4, align 16)
232B	  %18:fpr16 = FLH %fixed-stack.3, 0 :: (load (s16) from %fixed-stack.3, align 8)
240B	  %17:fpr32 = FLW %fixed-stack.2, 0 :: (load (s32) from %fixed-stack.2, align 16)
248B	  %16:fpr64 = FLD %fixed-stack.1, 0 :: (load (s64) from %fixed-stack.1)
256B	  FSH %19:fpr16, %23:gpr, 0 :: (volatile store (s16) into %ir.p)
264B	  FSH %18:fpr16, %23:gpr, 0 :: (volatile store (s16) into %ir.p)
272B	  FSW %17:fpr32, %23:gpr, 0 :: (volatile store (s32) into %ir.p)
288B	  FSD %16:fpr64, %23:gpr, 0 :: (volatile store (s64) into %ir.p)
304B	  PseudoRET
```

So the volatile stores are inserted there to prevent the loads from being sunk so we can ensure they are spilled. I'll try clarifying the comment

lukel97 committed:
Clarify why we need the volatile stores

lukel97 committed:
Merge branch 'main' into riscv/ld-remat

lukel97 commented:
Ping

lukel97 committed:
Change to external constant

lukel97 committed:
Adjust comment

lukel97 commented on llvm/test/CodeGen/RISCV/remat.ll:
```diff
@@ -200,3 +200,168 @@ for.end:                                          ; preds = %for.inc, %entry
+; We could remat the load of the constant global if we extended the live
+; interval of the high bits of the address.
+
+@const = constant i32 42
```
Good point, changed to an external constant in 86ddeeb8c24c574dcb38a22450d4eb6d1cc3a395. Looks like `EarlyCSEPass` handles the folding in the middle end, so these llc tests wouldn't pick it up.

preames reviewed (approved):
LGTM

lukel97 committed:
Merge branch 'main' into riscv/ld-remat
